OpenRA.Game/Server/PlayerMessageTracker.cs

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2007-2022 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+
+namespace OpenRA.Server
+{
+	class PlayerMessageTracker
+	{
+		[TranslationReference("remaining")]
+		static readonly string ChatDisabled = "chat-disabled";
+
+		readonly Dictionary<int, List<long>> messageTracker = new Dictionary<int, List<long>>();
+		readonly Server server;
+		readonly Action<Connection, int, int, byte[]> dispatchOrdersToClient;
+		readonly Action<Connection, string, Dictionary<string, object>> sendLocalizedMessageTo;
+
+		public PlayerMessageTracker(Server server, Action<Connection, int, int, byte[]> dispatchOrdersToClient, Action<Connection, string, Dictionary<string, object>> sendLocalizedMessageTo)
+		{
+			this.server = server;
+			this.dispatchOrdersToClient = dispatchOrdersToClient;
+			this.sendLocalizedMessageTo = sendLocalizedMessageTo;
+		}
+
+		public void DisableChat(Connection conn, int time)
+		{
+			dispatchOrdersToClient(conn, 0, 0, new Order("DisableChatEntry", null, false) { ExtraData = (uint)time }.Serialize());
+		}
+
+		public bool IsPlayerAtFloodLimit(Connection conn)
+		{
+			if (!messageTracker.ContainsKey(conn.PlayerIndex))
+				messageTracker.Add(conn.PlayerIndex, new List<long>());
+
+			var isAdmin = server.GetClient(conn)?.IsAdmin ?? false;
+			var settings = server.Settings;
+			var time = conn.ConnectionTimer.ElapsedMilliseconds;
+			var tracker = messageTracker[conn.PlayerIndex];
+			tracker.RemoveAll(t => t + settings.FloodLimitInterval < time);
+
+			long CalculateRemaining(long cooldown)
+			{
+				return (cooldown - time + 999) / 1000;
+			}
+
+			// Block messages untill join cooldown times out
+			if (!isAdmin && time < settings.FloodLimitJoinCooldown)
+			{
+				var remaining = CalculateRemaining(settings.FloodLimitJoinCooldown);
+				sendLocalizedMessageTo(conn, ChatDisabled, Translation.Arguments("remaining", remaining));
+				return true;
+			}
+
+			// Block messages if above flood limit
+			if (tracker.Count >= settings.FloodLimitMessageCount)
+			{
+				var remaining = CalculateRemaining(tracker[0] + settings.FloodLimitInterval);
+				sendLocalizedMessageTo(conn, ChatDisabled, Translation.Arguments("remaining", remaining));
+				return true;
+			}
+
+			tracker.Add(time);
+
+			// Disable chat UI when player has reached the flood limit
+			if (tracker.Count >= settings.FloodLimitMessageCount)
+			{
+				var cooldownDelta = Math.Max(0, settings.FloodLimitCooldown - settings.FloodLimitInterval);
+				for (var i = 0; i < tracker.Count; i++)
+					tracker[i] = time + cooldownDelta;
+
+				DisableChat(conn, settings.FloodLimitCooldown);
```

Comment:
I suggest
```suggestion
				// Disable the chat for future messages, but don't discard the current message by returning true
				DisableChat(conn, settings.FloodLimitCooldown);
```
to make clear this is intentional behaviour here. Otherwise someone in the future might add the return back, thinking they're fixing an oversight.

Reply:
I renamed the function to `DisableChatUI`. I don't think it's confusing to lack a return statement here. The above 2 statements `Block` messages, and this one just enables blocking, thus the name `Disable`